src/jitter.rs

```diff
@@ -0,0 +1,743 @@
+    // the loop in this function implies that careful retesting must be done.
+    #[inline(never)]
+    fn lfsr_time(&mut self, time: u64, var_rounds: bool) {
+        fn lfsr(mut data: u64, time: u64) -> u64{
+            for i in 1..65 {
+                let mut tmp = time << (64 - i);
+                tmp = tmp >> (64 - 1);
+
+                // Fibonacci LSFR with polynomial of
+                // x^64 + x^61 + x^56 + x^31 + x^28 + x^23 + 1 which is
+                // primitive according to
+                // http://poincare.matf.bg.ac.rs/~ezivkovm/publications/primpol1.pdf
+                // (the shift values are the polynomial values minus one
+                // due to counting bits from 0 to 63). As the current
+                // position is always the LSB, the polynomial only needs
+                // to shift data in from the left without wrap.
+                data ^= tmp;
+                data ^= (data >> 63) & 1;
+                data ^= (data >> 60) & 1;
+                data ^= (data >> 55) & 1;
+                data ^= (data >> 30) & 1;
+                data ^= (data >> 27) & 1;
+                data ^= (data >> 22) & 1;
+                data = data.rotate_left(1);
+            }
+            data
+        }
+
+        // Note: in the reference implementation only the last round effects
+        // `self.data`, all the other results are ignored. To make sure the
+        // other rounds are not optimised out, we first run all but the last
+        // round on a throw-away value instead of the real `self.data`.
+        let mut lfsr_loop_cnt = 0;
+        if var_rounds { lfsr_loop_cnt = self.random_loop_cnt(4) };
+
+        let mut throw_away: u64 = 0;
+        for _ in 0..lfsr_loop_cnt {
+            throw_away = lfsr(throw_away, time);
+        }
+        black_box(throw_away);
+
+        self.data = lfsr(self.data, time);
+    }
+
+    // Memory Access noise source
+    // This is a noise source based on variations in memory access times
+    //
+    // This function performs memory accesses which will add to the timing
+    // variations due to an unknown amount of CPU wait states that need to be
+    // added when accessing memory. The memory size should be larger than the L1
+    // caches as outlined in the documentation and the associated testing.
+    //
+    // The L1 cache has a very high bandwidth, albeit its access rate is usually
+    // slower than accessing CPU registers. Therefore, L1 accesses only add
+    // minimal variations as the CPU has hardly to wait. Starting with L2,
+    // significant variations are added because L2 typically does not belong to
+    // the CPU any more and therefore a wider range of CPU wait states is
+    // necessary for accesses. L3 and real memory accesses have even a wider
+    // range of wait states. However, to reliably access either L3 or memory,
+    // the `self.mem` memory must be quite large which is usually not desirable.

```

Comment:
Sounds to me like accesses in this block could be replaced with seqcst atomic operations and memory fences to make it contribute a significantly greater amount of jitter per round.

Reply:
But the RNG isn't multi-threaded? Also part of the goal was to make the code usable on embedded systems which provide their own timer (the timer itself is the only part of `JitterRng` which is not `no_std` compatible).

Reply:
Fences (a building block for atomic memory operations) are also available and important on single-threaded machines as well. For example (citing [this](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dai0321a/BIHJGFEE.html), but applicable in general):

* processors can re-order memory transfers to improve performance if it does not affect the behavior of the instruction sequence;
* processors can have multiple bus interfaces that have different wait-states;
* processors can have write buffers, and some write buffers can merge multiple transfers into a single transfer.

Only the most primitive chip designs will not have barriers and would fall back to regular reads, however these same most primitive designs have fairly constant memory access latency, and memory accesses would contribute no jitter anyway.

I don’t think I see any reason to not use what’s essentially the most jitter-susceptible and fairly well supported operation to collect entropy. 

---

At least I would be significantly more comfortable with implementation using fences than the one that tries to outsmart the modern processors’ caching strategy with what seems to be a linear access pattern.

Reply:
That seems like a better way! (I have never used barriers though).

Yet I would like to leave it as it is for now, because this mirrors jitterentropy. That one has research and statistics done on it, and it is nice to be able to point there as a 'proof' our entropy collection method works. That and the NIST entropy test.

Reply:
Note that JitterRng can work pretty well without this part, previous versions of jitterentropy did not have it. If in the future CPU's are to smart for this method `test_timer` will just say more rounds of `measure_jitter` are necessary .

Reply:
I’m fine with keeping the implementation as-is for now, but I would like to see an issue filled to collect information and explore better ways to collect jitter-based entropy.

http://chronox.de/jent/doc/CPU-Jitter-NPTRNG.html is something I’ve found a few minutes ago. It also references a number of other jitter-based RNG implementations.

Reply:
That is exactly the one this is based on. If you look at the original pr https://github.com/dhardy/rand/pull/28, I have a comment and commit messages so that our Rust code can be read side-by-side with the code from jitterentropy.